In [72]:
# #Set the enviroment variables
# import os
# os.environ['KAGGLE_USERNAME'] = "dlromanoff"
# os.environ['KAGGLE_KEY'] = "a63612e0bc2ceaee594bcec8d8f498b6"

In [1]:
# Imports needed or not
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder
from scipy import stats
import statsmodels.api as sm
import statsmodels.api as sm
import category_encoders as ce
from sklearn.preprocessing import scale
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV
from sklearn.feature_selection import f_classif, SelectKBest
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import PolynomialFeatures
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import mean_squared_error, r2_score
from mpl_toolkits.mplot3d import Axes3D

In [2]:
# Trying to see all the columns of the dataframe
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

In [3]:
# Read the csv's
train_features = pd.read_csv('train_features.csv')
test_features = pd.read_csv('test_features.csv')
train_labels = pd.read_csv('train_labels.csv')
train_features.head()

id  amount_tsh date_recorded        funder  gps_height     installer  longitude   latitude              wpt_name  num_private                    basin  subvillage   region  region_code  district_code        lga        ward  population public_meeting              recorded_by scheme_management                  scheme_name permit  construction_year extraction_type extraction_type_group extraction_type_class management management_group         payment payment_type water_quality quality_group      quantity quantity_group                source           source_type source_class              waterpoint_type waterpoint_type_group
0  69572      6000.0    2011-03-14         Roman        1390         Roman  34.938093  -9.856322                  none            0               Lake Nyasa    Mnyusi B   Iringa           11              5     Ludewa    Mundindi         109           True  GeoData Consultants Ltd               VWC                        Roman  False               1999         gravity               gravity               gravity        vwc       user-group    pay annually     annually          soft          good        enough         enough                spring                spring  groundwater           communal standpipe    communal standpipe
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI  34.698766  -2.147466              Zahanati            0            Lake Victoria     Nyamara     Mara           20              2  Serengeti       Natta         280            NaN  GeoData Consultants Ltd             Other                          NaN   True               2010         gravity               gravity               gravity        wug       user-group       never pay    never pay          soft          good  insufficient   insufficient  rainwater harvesting  rainwater harvesting      surface           communal standpipe    communal standpipe
2  34310        25.0    2013-02-25  Lottery Club         686  World vision  37.460664  -3.821329           Kwa Mahundi            0                  Pangani     Majengo  Manyara           21              4  Simanjiro     Ngorika         250           True  GeoData Consultants Ltd               VWC  Nyumba ya mungu pipe scheme   True               2009         gravity               gravity               gravity        vwc       user-group  pay per bucket   per bucket          soft          good        enough         enough                   dam                   dam      surface  communal standpipe multiple    communal standpipe
3  67743         0.0    2013-01-28        Unicef         263        UNICEF  38.486161 -11.155298  Zahanati Ya Nanyumbu            0  Ruvuma / Southern Coast  Mahakamani   Mtwara           90             63   Nanyumbu    Nanyumbu          58           True  GeoData Consultants Ltd               VWC                          NaN   True               1986     submersible           submersible           submersible        vwc       user-group       never pay    never pay          soft          good           dry            dry           machine dbh              borehole  groundwater  communal standpipe multiple    communal standpipe
4  19728         0.0    2011-07-13   Action In A           0       Artisan  31.130847  -1.825359               Shuleni            0            Lake Victoria  Kyanyamisa   Kagera           18              1    Karagwe  Nyakasimbi           0           True  GeoData Consultants Ltd               NaN                          NaN   True                  0         gravity               gravity               gravity      other            other       never pay    never pay          soft          good      seasonal       seasonal  rainwater harvesting  rainwater harvesting      surface           communal standpipe    communal standpipe

In [4]:
# Get some idea of the numbers I am working with
train_features.describe()

id     amount_tsh    gps_height     longitude      latitude   num_private   region_code  district_code    population  construction_year
count  59400.000000   59400.000000  59400.000000  59400.000000  5.940000e+04  59400.000000  59400.000000   59400.000000  59400.000000       59400.000000
mean   37115.131768     317.650385    668.297239     34.077427 -5.706033e+00      0.474141     15.297003       5.629747    179.909983        1300.652475
std    21453.128371    2997.574558    693.116350      6.567432  2.946019e+00     12.236230     17.587406       9.633649    471.482176         951.620547
min        0.000000       0.000000    -90.000000      0.000000 -1.164944e+01      0.000000      1.000000       0.000000      0.000000           0.000000
25%    18519.750000       0.000000      0.000000     33.090347 -8.540621e+00      0.000000      5.000000       2.000000      0.000000           0.000000
50%    37061.500000       0.000000    369.000000     34.908743 -5.021597e+00      0.000000     12.000000       3.000000     25.000000        1986.000000
75%    55656.500000      20.000000   1319.250000     37.178387 -3.326156e+00      0.000000     17.000000       5.000000    215.000000        2004.000000
max    74247.000000  350000.000000   2770.000000     40.345193 -2.000000e-08   1776.000000     99.000000      80.000000  30500.000000        2013.000000

In [5]:
# See what's going in the labels dataframe
train_labels.describe(include='object')

status_group
count         59400
unique            3
top      functional
freq          32259

Labels dataframe includes the status_group along with the id. The status group is the dependent variable here. The id needs to be used to track the status group.

In [6]:
# Checking for values
train_labels.status_group.value_counts(normalize=True)

functional                 0.543081
non functional             0.384242
functional needs repair    0.072677
Name: status_group, dtype: float64

In [7]:
y = train_labels.drop(columns='id')

**This indicates what our expectations should be for our outcomes. Function here is ~53.31%. Non-Function is ~38.42%. And, functional needs repair is ~7.27%**

In [8]:
X_train, X_test, y_train, y_test = train_test_split(
        train_features, y, train_size=0.75, test_size=0.25, random_state=42, shuffle=True)
    
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((44550, 40), (14850, 40), (44550, 1), (14850, 1))

In [9]:
# Baseline Prediction
majority_class = train_labels['status_group'].mode()[0]
y_pred = np.full(shape=train_labels['status_group'].shape, fill_value=majority_class)
print(train_labels.status_group.shape)
y_pred.shape

(59400,)


(59400,)

In [10]:
accuracy_score(train_labels['status_group'], y_pred)

0.543080808080808

In [11]:
# Making submission DataFrame
submission = pd.DataFrame(y_pred, columns=['status_group_included'])
submission = pd.concat([train_labels, submission], axis=1)
submission.head()

id    status_group status_group_included
0  69572      functional            functional
1   8776      functional            functional
2  34310      functional            functional
3  67743  non functional            functional
4  19728      functional            functional

In [12]:
submission = submission.drop(columns='status_group')
print(submission.shape)
submission.head()

(59400, 2)


id status_group_included
0  69572            functional
1   8776            functional
2  34310            functional
3  67743            functional
4  19728            functional

In [13]:
# Making all objects into numbers
def encode_cat(df):
        columnsToEncode = list(df.select_dtypes(include=['category','object']))
        le = LabelEncoder()
        for feature in columnsToEncode:
            try:
                df[feature] = le.fit_transform(df[feature])
            except:
                print('Error encoding '+feature)
        return df

In [14]:
# Checking for null values
train_features.isnull().sum()

id                           0
amount_tsh                   0
date_recorded                0
funder                    3635
gps_height                   0
installer                 3655
longitude                    0
latitude                     0
wpt_name                     0
num_private                  0
basin                        0
subvillage                 371
region                       0
region_code                  0
district_code                0
lga                          0
ward                         0
population                   0
public_meeting            3334
recorded_by                  0
scheme_management         3877
scheme_name              28166
permit                    3056
construction_year            0
extraction_type              0
extraction_type_group        0
extraction_type_class        0
management                   0
management_group             0
payment                      0
payment_type                 0
water_quality                0
quality_

In [15]:
# Filling null values with the mode
train_features['funder'] = train_features['funder'].fillna(train_features['funder'].mode()[0])
train_features['installer'] = train_features['installer'].fillna(train_features['installer'].mode()[0])
train_features['subvillage'] = train_features['subvillage'].fillna(train_features['subvillage'].mode()[0])
train_features['public_meeting'] = train_features['public_meeting'].fillna(train_features['public_meeting'].mode()[0])
train_features['scheme_management'] = train_features['scheme_management'].fillna(train_features['scheme_management'].mode()[0])
train_features['permit'] = train_features['permit'].fillna(train_features['permit'].mode()[0])
train_features['scheme_name'] = train_features['scheme_name'].fillna(train_features['scheme_name'].mode()[0])
# train_features.head()
# train_features.isnull().sum()

In [16]:
test_features.isnull().sum()

id                          0
amount_tsh                  0
date_recorded               0
funder                    783
gps_height                  0
installer                 788
longitude                   0
latitude                    0
wpt_name                    0
num_private                 0
basin                       0
subvillage                 94
region                      0
region_code                 0
district_code               0
lga                         0
ward                        0
population                  0
public_meeting            785
recorded_by                 0
scheme_management         939
scheme_name              6839
permit                    663
construction_year           0
extraction_type             0
extraction_type_group       0
extraction_type_class       0
management                  0
management_group            0
payment                     0
payment_type                0
water_quality               0
quality_group               0
quantity  

In [17]:
X_test = test_features.copy()
X_test.shape

(14358, 40)

In [18]:
X_test['funder'] = test_features['funder'].fillna(test_features['funder'].mode()[0])
X_test['installer'] = test_features['installer'].fillna(test_features['installer'].mode()[0])
X_test['subvillage'] = test_features['subvillage'].fillna(test_features['subvillage'].mode()[0])
X_test['public_meeting'] = test_features['public_meeting'].fillna(test_features['public_meeting'].mode()[0])
X_test['scheme_management'] = test_features['scheme_management'].fillna(test_features['scheme_management'].mode()[0])
X_test['permit'] = test_features['permit'].fillna(test_features['permit'].mode()[0])
X_test['scheme_name'] = test_features['scheme_name'].fillna(test_features['scheme_name'].mode()[0])

In [19]:
X_test.isnull().sum()

id                       0
amount_tsh               0
date_recorded            0
funder                   0
gps_height               0
installer                0
longitude                0
latitude                 0
wpt_name                 0
num_private              0
basin                    0
subvillage               0
region                   0
region_code              0
district_code            0
lga                      0
ward                     0
population               0
public_meeting           0
recorded_by              0
scheme_management        0
scheme_name              0
permit                   0
construction_year        0
extraction_type          0
extraction_type_group    0
extraction_type_class    0
management               0
management_group         0
payment                  0
payment_type             0
water_quality            0
quality_group            0
quantity                 0
quantity_group           0
source                   0
source_type              0
s

In [20]:
X_test.shape

(14358, 40)

In [21]:
# Baseline Prediction
# majority_class = y_train.mode()
# y_pred = np.full(shape=y_test.shape, fill_value=majority_class)
# print(train_labels.status_group.shape)
# y_pred.shape

In [22]:
# accuracy_score(y_test, y_pred)

In [23]:
# Using only numeric columns for now
X = train_features.select_dtypes('number')
X = X.drop(columns='id')
X = scale(X)

/Users/danielleromanoff/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by the scale function.
  after removing the cwd from sys.path.


In [24]:
X_train, X_test, y_train, y_test = train_test_split(
        X, y, train_size=0.75, test_size=0.25, random_state=42, shuffle=True)
    
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((44550, 9), (14850, 9), (44550, 1), (14850, 1))

In [25]:
logreg = LogisticRegression()
logreg.fit(X_train,y_train)
y_pred = logreg.predict(X_train)
accuracy_score(y_train,y_pred)

/Users/danielleromanoff/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/danielleromanoff/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/danielleromanoff/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


0.55672278338945

In [28]:
# Using test data
X_test_data = X_test.drop(columns='id')
X_test_data = X_test_data.select_dtypes('number')
X_test_data = scale(X_test_data)
#Testing with Test data
X_test_data

AttributeError: 'numpy.ndarray' object has no attribute 'drop'

In [487]:
# Logistic Regression Model
y_pred = logreg.predict(X_test_data)
prediction_df = pd.DataFrame(y_pred)
identities = pd.DataFrame(X_test['id'])
print(identities.shape)
test_first_pass = identities.join(prediction_df)
test_first_pass.rename(columns={0: 'status_group'}, inplace=True)

(14358, 1)


In [494]:
print(test_first_pass.shape)
test_first_pass.head()

(14358, 2)


id    status_group
0  50785      functional
1  51630      functional
2  17168      functional
3  45559  non functional
4  49871      functional

In [497]:
test_first_pass.to_csv(path_or_buf="/Users/danielleromanoff/Documents/Lambda/DS-Unit-2-Project-Week/Submission.csv", index=False)

In [491]:
# print('Classification Report')
# print()
# print(classification_report(X_test_data, y_pred))
# print('The accuracy is: ', accuracy_score(X_test_data, y_pred))

In [492]:
test_first_pass.set_index('id')

status_group
id                   
50785      functional
51630      functional
17168      functional
45559  non functional
49871      functional
52449      functional
24806      functional
28965      functional
36301  non functional
54122      functional
419        functional
45750      functional
653        functional
14017      functional
40228      functional
27714      functional
28785      functional
18532      functional
69961      functional
55083      functional
8691       functional
30331      functional
70970      functional
61136      functional
28799      functional
46825      functional
44718      functional
37350      functional
65738      functional
3271       functional
42658      functional
67523      functional
37034      functional
42432      functional
27420      functional
707        functional
48489      functional
51008      functional
72562  non functional
10275      functional
53540      functional
43291      functional
62343      functional
1649       functional
8822       functional
10352      functional
50878      functional
15266      functional
51956      functional
44099      functional
59         functional
51359  non functional
15962      functional
17853      functional
42040      functional
72243      functional
63653      functional
72053      functional
29351      functional
51897      functional
20376      functional
26804      functional
67303      functional
69780      functional
54665      functional
39341      functional
8869       functional
20691      functional
19693      functional
14321      functional
63059      functional
67403      functional
28912      functional
12549      functional
24328      functional
13750      functional
1574       functional
44149      functional
53917      functional
22705      functional
15644      functional
20773      functional
33092      functional
54392      functional
37850      functional
5576       functional
9971       functional
57508      functional
25872      functional
43978      functional
38181      functional
24325      functional
2724       functional
34794      functional
56525      functional
63155      functional
67247      functional
23387  non functional
17873      functional
18197      functional
24834      functional
4091       functional
33261      functional
44384      functional
62096      functional
64554      functional
68314      functional
53         functional
23426      functional
38259      functional
16631      functional
43162      functional
44492      functional
38884      functional
13690      functional
60335      functional
31856      functional
67891      functional
25564      functional
55293      functional
31866      functional
71178      functional
60279      functional
28615      functional
8744       functional
48308      functional
49626      functional
38098      functional
30309      functional
8015       functional
3496       functional
56971      functional
68616      functional
8646       functional
74104      functional
37501  non functional
69202      functional
68409      functional
51847      functional
55647      functional
17923      functional
71743      functional
51539      functional
73819      functional
66881      functional
58768      functional
2648       functional
68933      functional
12427      functional
16633      functional
24123      functional
58551      functional
17132      functional
3027       functional
2651       functional
25998      functional
63400      functional
50028      functional
67897      functional
21425      functional
71756      functional
16862      functional
15420      functional
712    non functional
54368      functional
2866       functional
51130      functional
27993      functional
21621      functional
66038      functional
42182      functional
17694  non functional
57759      functional
47978      functional
52359      functional
69656      functional
24086      functional
43579      functional
43284      functional
2877       functional
21171

In [282]:
prediction_second = pd.DataFrame(X_test.id.values, columns=['id'])
prediction_second.shape

(14850, 1)

In [283]:
# Combining the label from the numeric dataframe and the prediction
prediction = pd.concat([prediction_second, prediction], axis=1)
prediction.columns.values[1]= 'status_group'
# prediction.head()
prediction.shape

(14850, 2)

In [284]:
prediction.to_csv(path_or_buf="/Users/danielleromanoff/Documents/Lambda/DS-Unit-2-Project-Week/Prediction_1.csv", sep=',', index=False)

In [175]:
# Finding null values after the merge.
train_new.isnull().sum()

id                           0
status_group                 0
amount_tsh                   0
date_recorded                0
funder                    3635
gps_height                   0
installer                 3655
longitude                    0
latitude                     0
wpt_name                     0
num_private                  0
basin                        0
subvillage                 371
region                       0
region_code                  0
district_code                0
lga                          0
ward                         0
population                   0
public_meeting            3334
recorded_by                  0
scheme_management         3877
scheme_name              28166
permit                    3056
construction_year            0
extraction_type              0
extraction_type_group        0
extraction_type_class        0
management                   0
management_group             0
payment                      0
payment_type                 0
water_qu

In [149]:
# For now just dropping NaN's and checking to make sure there are none left
# train_clean = train_new.dropna(axis=1)
# train_clean.isnull().sum()

In [150]:
# Making a list of the columns in the new df
# print(list(train_clean))

In [94]:
# Assigning X and y values for Logistic Regression
# X = train_clean.drop(['status_group'], axis=1)
# y = train_clean['status_group']

In [113]:
# Encoding X_train and X_test
X_train = encode_cat(X_train)
X_test = encode_cat(X_test)
print('X_train shape is ',X_train.shape)
print('X_test shape is ',X_test.shape)

X_train shape is  (44550, 33)
X_test shape is  (14850, 33)


In [116]:
X_train.head()

id  amount_tsh  date_recorded  gps_height  longitude      latitude  wpt_name  num_private  basin  region  region_code  district_code  lga  ward  population  recorded_by  construction_year  extraction_type  extraction_type_group  extraction_type_class  management  management_group  payment  payment_type  water_quality  quality_group  quantity  quantity_group  source  source_type  source_class  waterpoint_type  waterpoint_type_group
24947  33935        20.0             41         330  38.123839 -6.087137e+00      1513            0      8      14            6              1    3  1172          80            0               1972                6                     10                      5           7                 4        4             5              4              4         2               2       3            0             0                1                      1
22630  49654         0.0            291           0   0.000000 -2.000000e-08     30655            0      4      13           17              1   54  1660           0            0                  0                8                      5                      1          11                 4        6             6              6              2         1               1       7            5             0                4                      3
13789  39287         0.0            135           0  33.312321 -2.814100e+00      1698            0      4      13           19              4   45  2070           0            0                  0                4                      2                      1           7                 4        0             2              6              2         1               1       3            0             0                4                      3
15697  60510         0.0            264        1542  34.783049 -4.842093e+00     37195            0      0      18           13              4  110  1428          23            0               2010                3                      1                      0          10                 4        0             2              6              2         3               3       5            3             1                1                      1
22613  24259         0.0            348         523  34.660944 -1.070733e+01     17901            0      1      16           10              3   61  1139           1            0               2013                3                      1                      0           7                 4        6             6              6              2         0               0       8            6             0                1                      1

In [117]:
# Fitting the model and getting its accuracy score
model= LogisticRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
accuracy_score(y_test, y_pred)

/Users/danielleromanoff/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/danielleromanoff/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


0.635959595959596

In [125]:
# Using pipeline with the training data
pipeline = make_pipeline(ce.OneHotEncoder(use_cat_names=True),
                         StandardScaler(), LogisticRegression(solver ='lbfgs',n_jobs=-1, multi_class = 'auto',C=2))
pipeline.fit(X_train, y_train)

/Users/danielleromanoff/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/Users/danielleromanoff/anaconda3/lib/python3.6/site-packages/sklearn/base.py:465: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, y, **fit_params).transform(X)


Pipeline(memory=None,
     steps=[('onehotencoder', OneHotEncoder(cols=[], drop_invariant=False, handle_unknown='impute',
       impute_missing=True, return_df=True, use_cat_names=True, verbose=0)), ('standardscaler', StandardScaler(copy=True, with_mean=True, with_std=True)), ('logisticregression', LogisticRegression(C=2, cla...enalty='l2', random_state=None, solver='lbfgs', tol=0.0001,
          verbose=0, warm_start=False))])

In [128]:
# y prediction using the pipeline
y_pred = pipeline.predict(X_train)

/Users/danielleromanoff/anaconda3/lib/python3.6/site-packages/sklearn/pipeline.py:331: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  Xt = transform.transform(Xt)


In [131]:
# Putting results in a dataframe
prediction = pd.DataFrame(y_pred, X_train['id'])
prediction.columns = ['status_group']
print(prediction.shape)
prediction.head()

(44550, 1)


status_group
id                   
33935      functional
49654      functional
39287  non functional
60510      functional
24259  non functional

In [174]:
prediction.to_csv(path_or_buf="/Users/danielleromanoff/Documents/Lambda/DS-Unit-2-Project-Week/Prediction_1.csv", sep=',', index=False)

In [143]:
# Making the 'status_group' a column
fix_df= pd.DataFrame(prediction)
submit = fix_df.index
submission_1 = pd.DataFrame(fix_df['status_group'],submit)
submission_1.reset_index(inplace = True)
submission_1.head()

id    status_group
0  33935      functional
1  49654      functional
2  39287  non functional
3  60510      functional
4  24259  non functional

In [144]:
# Dropping all status_group values that are not equal to 'functional'
submission_1 = submission_1[submission_1["status_group"] == 'functional']
print(submission_1.shape)
submission_1.head()

(31302, 2)


id status_group
0  33935   functional
1  49654   functional
3  60510   functional
5  52400   functional
7   3435   functional

In [146]:
# newsub = pd.DataFrame(submission_1)
# print(newsub.shape)
# sub_2 = newsub.index
# submission_1 = pd.DataFrame(newsub['status_group'], sub_2)
# print(submission_1.head())
# subm.reset_index(inplace = True)

In [61]:
submission_1.to_csv(path_or_buf="/Users/danielleromanoff/Documents/Lambda/DS-Unit-2-Project-Week/Submission_1.csv", index=False)

In [ ]:
# Assigning X and y values for Logistic Regression
X = train_new.drop(['status_group'], axis=1)
y = train_new['status_group']